M2351: Enhance secure/non-secure image build flow for non-PSA target

ccli8 opened the pull request:
### Description

This PR is to enhance secure/non-secure image build flow for non-PSA target. Its update is abstracted from #10959 just for the non-PSA part to improve build UX. It is expected that this PR and #10959 are compatible for the build flow. It has the following modifications:

1. Rename non-PSA target name to **NUMAKER_PFM_M2351_NOPSA_S/NS**
1. Support non-PSA secure/non-secure combined build

The above modifications enhance build flow but at the same time cause incompatibility with current one:
1. The target name **NUMAKER_PFM_M2351** is recycled. User needs to change to **NUMAKER_PFM_M2351_NOPSA_S/NS** to build non-PSA secure/non-secure targets.
1. Just flash combined secure/non-secure image altogether instead of flash secure image first and then non-secure image.
1.  User can also drop pre-built secure image saved in `TARGET_NU_PREBUILD_SECURE` directory and provide its own by adding the line below in **mbed_app.json** instead of removing it through **.mbedignore**:
```
    "target.extra_labels_remove": ["NU_PREBUILD_SECURE"]
```

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [X] Breaking change

### Release Notes

Changes for NuMaker-PFM-M2351:
1. Change target name to NU_PFM_M2351_NPSA_S for non-PSA secure target and and NU_PFM_M2351_NPSA_NS for non-PSA non-secure target. Original target name NUMAKER_PFM_M2351 is recycled.
2. Combine secure image and non-secure image into one in non-secure target post-build. With this, user just needs to flash the combined image instead of flash secure image and then non-secure image separately.
3. To drop pre-built secure image in mbed-os tree and provide custom one, add the line '"target.extra_labels_remove": ["NU_PREBUILD_SECURE"]' into 'mbed_app.json' instead of via .mbedignnore.

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

ccli8 committed:
M2351: Remove pre-built non-PSA secure image temporarily

This will add back immediately after target renaming is done.

ccli8 committed:
M2351: Rename non-PSA target name to NU_PFM_M2351_NPSA_S/NS

1.  Create a private target name NU_PFM_M2351_CM which stands for the
    NuMaker-PFM-M2351 board and is to be extended.
2.  NU_PFM_M2351_NPSA_S/NS target names for non-PSA secure/non-secure targets
    respectively.
3.  The original target name NUMAKER_PFM_M2351 is recycled and cannot be used.
    Use NU_PFM_M2351_S/NS for non-PSA secure/non-secure targets instead.

NOTE:   Target name doesn't follow the rule below because online database has
        limit of max 20 chars:

        NUMAKER_PFM_M2351_PSA/NOPSA_S/NS

        Instead, it has the rule:

        NU_PFM_M2351_[NPSA_]S/NS

        NU_PFM_M2351_S/NS for PSA targets. This is to be consistent with current
        PSA target naming. So the resolved target names are:

        NU_PFM_M2351_S          : PSA secure target
        NU_PFM_M2351_NS         : PSA non-secure target
        NU_PFM_M2351_NPSA_S     : Non-PSA secure target
        NU_PFM_M2351_NPSA_NS    : Non-PSA non-secure target

ccli8 committed:
M2351: Support non-PSA secure/non-secure combined build

Support secure/non-secure combined build for non-PSA target:
1.  In secure post-build, deliver built secure image to TARGET_NU_PREBUILD_SECURE
    directory which is to combine later.
2.  In non-secure post-build, merge non-secure image with secure image saved in
    TARGET_NU_PREBUILD_SECURE directory.
3.  In non-secure post-build, user can also drop pre-built secure image saved in
    TARGET_NU_PREBUILD_SECURE directory and provide its own by adding the line below
    in mbed_app.json:
    "target.extra_labels_remove": ["NU_PREBUILD_SECURE"]

madchutney commented on targets/targets.json:
```diff
@@ -8353,8 +8353,7 @@
     },
-    "NUMAKER_PFM_M2351": {
-        "core": "Cortex-M23-NS",
+    "NUMAKER_PFM_M2351_CM": {
```
This change needs to be replicated into the online database (please ask you TAM representative) and the platform database in https://github.com/ARMmbed/mbed-os-tools

ccli8 replied:
> This change needs to be replicated into the online database (please ask you TAM representative)

OK

> the platform database in https://github.com/ARMmbed/mbed-os-tools

The platform name **NUMAKER_PFM_M2351** in mbed-os-tools database is kept unchanged for compatibility. Don't want to meet mismatch issue with mbed-os-tools and mbed-os.

madchutney replied:
I'm not sure what your comments mean, can you explain? The names need the same in all places or the entire build system, testing and reporting will not work.

mbed-os-tools defines the board type when the product/detect code from the board is detected. This means that some systems will attempts to build based on this board type, which will not be present in mbed-os.

ccli8 replied:
@madchutney Originally, I want to keep the [M2351 platform name in mbed-os-tools](https://github.com/ARMmbed/mbed-os-tools/blob/e732afb07a8bfc3796a522074c4ae4ad27b19637/src/mbed_os_tools/detect/platform_database.py#L227) unchanged, so users don't need to care about version match between mbed-os and mbed-os-tools. Now that the name must be consistent across different tools, I will send another PR to mbed-os-tools repo to change platform name to **NU_PFM_M2351** from **NUMAKER_PFM_M2351** after this PR is confirmed. https://github.com/ARMmbed/mbed-os/pull/11288/commits/bcf8cc63b581d03be0d54ec4410934abdccb16af is addressing the target name/platform name conversion for M2351 like other PSA targets.

madchutney commented on targets/targets.json:
```diff
@@ -8420,17 +8419,52 @@
-        "mbed_ram_size"     : "0x10000"
+        "public": false
+    },
+    "NUMAKER_PFM_M2351_NOPSA_NS": {
```
This name is too long as it also needs to be in the online database (max 20 characters)

ccli8 replied:
Fixed

madchutney commented on targets/targets.json:
```diff
@@ -8420,17 +8419,52 @@
+        "mbed_ram_start"        : "0x30008000",
+        "mbed_ram_size"         : "0x10000"
+    },
+    "NUMAKER_PFM_M2351_NOPSA_S": {
```
This name is too long as it also needs to be in the online database (max 20 characters)

ccli8 replied:
Fixed

madchutney commented on tools/targets/__init__.py:
```diff
@@ -707,6 +707,46 @@ def binary_hook(t_self, resources, elf, binf):
+        assert os.path.isfile(ns_hex), "Non-secure image %s must be regular file" % s_hex
+
+        # Keep original non-secure before merge with secure
+        ns_nosecure_hex = _ + "_no-secure-merge" + ext
```
Please don't use `_` as variable name, convention says this can be used when a return value is not required but then to actually use it is wrong.

ccli8 replied:
Fixed

madchutney commented on tools/targets/__init__.py:
```diff
@@ -707,6 +707,46 @@ def binary_hook(t_self, resources, elf, binf):
+        t_self.notify.info("Found secure image %s" % s_hex)
+        
+        _, ext = os.path.splitext(s_hex)
+        assert ext == ".hex", "Secure image %s must be in Intel HEX format" % s_hex
```
This assert does not seem to be checking for a programming assertion but rather a test for the file retrieved.

ccli8 replied:
Fixed

mark-edgeworth commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_NU_PFM_M2351_NP_NS/TARGET_NU_PREBUILD_SECURE/NuMaker-mbed-TZ-secure-example.hex:
```diff
@@ -0,0 +1,1806 @@
+:020000040000FA
```
Should this file be checked in at all. What does it contain? Can you document the commands used to generate it and use these instead?

ccli8 replied:
@mark-edgeworth Added their purpose. For build process, user needs to follow the link documented in **README.md**.

mark-edgeworth commented on targets/targets.json:
```diff
@@ -8422,17 +8421,52 @@
             "MPU"
         ],
-        "components_add": ["FLASHIAP"],
+        "components_add": [],
```
Looks like this key should be removed altogether if empty. There appear to be no other examples of this key being used with an empty list.

mark-edgeworth commented on targets/targets.json:
```diff
@@ -8422,17 +8421,52 @@
+            "M23_NS",
+            "NU_PREBUILD_SECURE"
+        ],
+        "device_has_remove": [],
```
Remove this line

mark-edgeworth commented on targets/targets.json:
```diff
@@ -8422,17 +8421,52 @@
+            "MBED_TZ_DEFAULT_ACCESS=1"
+        ],
+        "components_add": ["FLASHIAP"],
+        "components_remove": [],
```
Remove this line too.

mark-edgeworth commented on targets/targets.json:
```diff
@@ -8422,17 +8421,52 @@
+        ],
+        "device_has_remove": ["TRNG", "SERIAL", "SERIAL_ASYNCH", "SERIAL_FC", "STDIO_MESSAGES"],
+        "components_add": ["FLASHIAP"],
+        "components_remove": [],
```
Remove this line

ccli8 replied:
@mark-edgeworth Removed empty list here and all above

mark-edgeworth reviewed (dismissed):
See inline for requested changes

mark-edgeworth commented on tools/test_api.py:
```diff
@@ -2272,7 +2272,17 @@ def build_tests(tests, base_source_paths, build_path, target, toolchain_name,
+            #    SUFFIX for this board.
+            # 2. The registered platform name 'NUMAKER_PFM_M2351' in mbed-os-tools
+            #    keeps unchanged for compatibility.
+            test_builds[build]['platform'] = 'NUMAKER_PFM_M2351'
```
I'm sorry, but we cannot accept this as it changes the ID. This is likely to have serious implications throughout the build system (apart from making it very difficult to debug and maintain).

ccli8 replied:
@mark-edgeworth OK, and I still need to update platform database in mbed-os-tools to change platform name to **NU_PFM_M2351** from **NUMAKER_PFM_M2351**. Note there're multiple target names for the same M2351 board due to combinations of PSA/non-PSA and secure/non-secure as the pattern **NU_PFM_M2351_[P, NP]_[S, NS]**, stripping suffix of these target names to **NU_PFM_M2351** is still inevitable.

mark-edgeworth replied:
Looks ok now; names fit in 20 characters.

0xc0170 commented on tools/test_api.py:
```diff
@@ -2272,7 +2272,18 @@ def build_tests(tests, base_source_paths, build_path, target, toolchain_name,
+            #    so we cannot follow the naming rule below for this board:
+            #    'NUMAKER_PFM_M2351'+'_'+['PSA', 'NOPSA']+'_'+['S', 'NS']
+            # 2. The originally registered platform name 'NUMAKER_PFM_M2351'
+            #    in mbed-os-tools must change to 'NU_PFM_M2351' accordingly.
```
@ARMmbed/mbed-os-tools  Do we need this hack, can't change the limit in the online database (wonder why is it only 20 char) ?

mark-edgeworth replied:
Fundamentally we do not know what the knock-on effect of increasing the database limit might be. The limit has been in place for several years: is there any code that depends on this? The easiest way to resolve this is to keep the board name to 20 characters or fewer, rather than introducing another hack...

0xc0170 replied:
@ccli8 What can we do about this hack?

mark-edgeworth replied:
Looks like the names have now been reduced to fit inside 20 characters, so now ok from our point of view. However, this code is owned by @jamesbeyond so he might well have an opinion on this.

jamesbeyond replied:
Technically this code is not own by me, because this is part of mbed-cli test interface  😉  @mark-edgeworth 

Though, seems the target name already been shortened, I can't see any reason, we need this hard-coding `if` statement to truncate the name, please confirm @ccli8.

ccli8 replied:
@jamesbeyond For M2351, the target name is **NU_PFM_M2351_NP_NS** for non-PSA non-secure target or future **NU_PFM_M2351_P_NS** for PSA non-secure target, but the platform name registered in mbed-os-tools is [**NU_PFM_M2351**](https://github.com/ARMmbed/mbed-os-tools/pull/196). The truncation here is to convert target name to mbed-os-tools platform name for Greentea test (written into test spec file). PSA target like **LPC55S69_NS** also has the same truncation.

<pre>
        elif Target.get_target(test_builds[build]['platform']).is_PSA_non_secure_target:
            <b>
            if test_builds[build]['platform'].endswith('_NS'):
                test_builds[build]['platform'] = test_builds[build]['platform'][:-3]
            </b>
            if test_builds[build]['platform'].endswith('_PSA'):
                test_builds[build]['platform'] = test_builds[build]['platform'][:-4]
</pre>

jamesbeyond replied:
Hi @ccli8, I believe the code that you quoting can remove the suffix of `_NS` from your targets. right? so you don't need to worried about the `_NS`. Then how about `_P` or `_NP`? what are they stand for?
In this script we generally not recommend to have code for specific target.  if the `_P` or `_NP` are something general, you can add a line to remove `_P` or `_NP` as well

ccli8 replied:
@jamesbeyond Due to max 20 chars limited in online database, `_P` for `PSA` and `_NP` for `_NOPSA`. `_P`/`_NP` are shortened names for M2351 and so target specific code here

jamesbeyond replied:
@ccli8 In this case, `NU_PFM_M2351_PSA_NS` and `NU_PFM_M2351_NS` should be the names for psa and non-psa variants, following the convention, they both less than 20 char. and they will be handled without extra truncating.

BTW, @MarceloSalazar, do we have naming guidelines for Partners porting their target to mbed-os? to let people know the limitation of 20 chars, and to avoid changing name in the future.

ccli8 replied:
@jamesbeyond OK. I will rename them to NU_PFM_M2351_[S, NS] and NU_PFM_M2351_PSA_[S, NS] for non-PSA target and PSA target respectively.

MarceloSalazar replied:
We don't have explicit guidelines, but usually recommend to choose a relatively small number of characters to help developers when using the platform, especially with Mbed CLI.
We shouldn't be changing the name at all and this should be considered an exception.

ccli8 replied:
@jamesbeyond Split out #11467 to address the name truncation for TZ targets. It is general and no target-specific.

0xc0170 reviewed (commented):
@ccli8 how to migrate current users to new naming? Is there a way if I use the current name on master and take this PR , I would compile without issues?

0xc0170 reviewed (approved):
Besides the hack in test_api. Not needed anymore?

ccli8 committed:
M2351: Enable USB UART for non-PSA secure target for Greentea

NOTE1:  USB UART is partitioned for non-secure world. Secure world still can share
        it with limit that its interrupt cannot use in secure world.
NOTE2:  In secure world, USB UART is only for Greentea and STDIO. Developers shouldn't
        use it for other purposes.

ccli8 committed:
M2351: Add post-binary hook into uvision whitelist

This is necessary for exporting M2351 uvision project.

urutva commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_NU_PFM_M2351_NS/TARGET_NU_PREBUILD_SECURE/NuMaker-mbed-TZ-secure-example.hex:
```diff
@@ -0,0 +1,1808 @@
+:020000040000FA
```
The location for this file should be `mbed-os/targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_ NU_PFM_M2351_NP_NS\TARGET_NU_PREBUILD_SECURE\NuMaker-mbed-TZ-secure-example.hex` as the location `TARGET_NU_PFM_M2351_NS` should be reserved for TF-M secure binary.

urutva commented on targets/TARGET_NUVOTON/TARGET_M2351/TARGET_NUMAKER_PFM_M2351/TARGET_M23_NS/README.md:
```diff
@@ -1,30 +0,0 @@
-# Default secure code for M2351
```
The documentation is informative. Can we retain this but move it under `mbed-os/targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_ NU_PFM_M2351_NP_NS`?

ccli8 replied:
@devran01 Add back missing license and readme files into **mbed-os/targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_NU_PFM_M2351_NS/TARGET_NU_PREBUILD_SECURE/**

urutva commented on targets/targets.json:
```diff
@@ -8730,17 +8729,48 @@
-        "mbed_ram_size"     : "0x10000"
+        "public": false
+    },
+    "NU_PFM_M2351_NS": {
```
The hex file added in this PR `NuMaker-mbed-TZ-secure-example.hex` indicates that it is for non-PSA target. 
According to https://github.com/ARMmbed/mbed-os/pull/11288#issuecomment-525657587,
The non-PSA target should be named `NU_PFM_M2351_NP_S ` and `NU_PFM_M2351_NP_NS `. What is the reason to add PSA specific targets?

urutva commented on targets/targets.json:
```diff
@@ -8730,17 +8729,48 @@
+        "mbed_ram_start"        : "0x30008000",
+        "mbed_ram_size"         : "0x10000"
+    },
+    "NU_PFM_M2351_S": {
```
The hex file added in this PR `NuMaker-mbed-TZ-secure-example.hex` indicates that it is for non-PSA target. 
According to https://github.com/ARMmbed/mbed-os/pull/11288#issuecomment-525657587,
The non-PSA target should be named `NU_PFM_M2351_NP_S ` and `NU_PFM_M2351_NP_NS `. What is the reason to add PSA specific targets?

urutva commented on targets/targets.json:
```diff
@@ -8730,17 +8729,48 @@
+        "device_has_remove": ["TRNG"],
+        "components_add": ["FLASHIAP"],
+        "deliver_to_target": "NU_PFM_M2351_NS",
+        "delivery_dir": "TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_NU_PFM_M2351_NS/TARGET_NU_PREBUILD_SECURE",
```
"delivery_dir" should be "TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_NU_PFM_M2351_NP_NS/TARGET_NU_PREBUILD_SECURE"

ccli8 replied:
@devran01 answer target renaming for all above. There is **another** target renaming for https://github.com/ARMmbed/mbed-os/pull/11288#discussion_r323304549 and #11467. The final resolved M2351 target names are:

- NU_PFM_M2351_PSA_S: Target name for building M2351 TFM secure code
- NU_PFM_M2351_PSA_NS: Target name for building M2351 TFM non-secure code
- NU_PFM_M2351_S: Target name for building M2351 non-TFM secure code
- NU_PFM_M2351_NS: Target name for building M2351 non-TFM non-secure code

So needn't fix for above comments.

#10959 will be updated after this PR is confirmd.

urutva replied:
@ccli8 Currently in mbed-os, PSA targets are named as targetname_S and targetname_NS for secure and non-secure respectively. Unfortunately, the proposed names are not following that. Due to that, I suggest the following,

```
NU_PFM_M2351_S: Target name for building M2351 TFM secure code
NU_PFM_M2351_NS: Target name for building M2351 TFM non-secure code
NU_PFM_M2351_NPSA_S: Target name for building M2351 non-TFM secure code
NU_PFM_M2351_NPSA_NS: Target name for building M2351 non-TFM non-secure code
```

Can you please update the target names and the respective folders?

ccli8 replied:
@devran01 It concerns something. First, #11467 doesn't support **NPSA** suffix. Second, **NU_PFM_M2351_NPSA_NS** is exactly 20 chars. Not sure if it breaks online database max 20 chars limit. If possible, prefer keeping the current naming.

urutva replied:
@ccli8 Thanks for highlighting potential issue with new naming and #11467. We need to fix that as well. 
But before that, @mark-edgeworth Is it ok to have target names with exactly 20 characters or should they be less than 20?

Unfortunately we cannot retain the current names as this will cause confusion with other existing PSA/TFM targets (https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L5678, https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L5700, https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L2115, https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L2168)

ccli8 replied:
@devran01 OK. I will rename as suggested.

mark-edgeworth replied:
There is at least one other name in the database that has 20 characters, so this should be ok.

ccli8 committed:
M2351: Add pre-built secure image for non-PSA
